### DISCUSSION TITLE
CEC Password Provisioning

### DISCUSSION THREAD

**Discussion by nieuwsma** [Cray-HPE/community#22] [CSM-Hardware Management SIG]
# Context
Today (2021Q3), Mountain hardware comes with a default root credential baked into the firmware image for the nC, CMM, sC. 

This is a concern for several reasons: 

1. regulatory
    2. [California Senate Bill 327](https://leginfo.legislature.ca.gov/faces/billNavClient.xhtml?bill_id=201720180SB327) - more or less requires that devices not have default passwords.   Basically you have to have unique passwords for each device or have to have a mechanism to force the user to set the password before first use.
1. default firmware passwords are a security antipattern, per [US CERT](https://us-cert.cisa.gov/ncas/alerts/TA13-175A)

## CEC Password Provisioning Process
There is a process built into the CEC (Cabinet Environmental Controller) to provision passwords; but CSM is not architecturally well-aligned with it. 

A CEC is an embedded linux RTOS that controls all the CMMs, nC (node controllers), and sC (switch controllers) for a cabinet.  

The (approximate) high level flow to provision the CEC password is to:

1. login to the CEC
    1. This can be done directly to the CEC via IPV6 
    2. or via the CMM over TELNET
1. Put the CEC into 'privileged' mode and request a one-time-use unlock token.  The token will display on the LCD panel.
2. Input the unlock code into the CEC and access the provisioning features.
3. Set the default password hash

Some relevant rules:

- The provisioning process will set the root linux password and redfish password on all devices below the CEC, this is not a process manufacturing can do on the behalf of the customer, but is a customer process; the customer must record the password.  
- Any 'new' devices that have not had their passwords set by a user will also get this same default credential.  Once a user sets the credential then the CEC will not overwrite it, unless ->
- An NVME wipe of a blade will cause the blade to be restored to the CEC default credential, rather than the firmware default credential. 
- An administrator must separately set all CEC credentials, there is no way to multi-cast this credential. 
- A human must be involved to read the one-time-use unlock code off the LCD panel.

## CSM Requirements
CSM has to know what the default or current credentials of the hardware is.  If an out-of-band mechanism, like the CEC provisioning tool is used, some mechanism or process would have to inform CSM component of the expected default CEC credentials.  This is very important in the context of discovery as we currently expect a default (that is stored in an encrypted sealed secrets cache.
Currently the manufacturing hard coded default password is loaded automatically into vault (via `customizations.yaml` -> encrypted). The CSM services have an ABSOLUTE dependency on knowing what the password a priori.

### Limitations

1.  CSM currently cannot use any IPV6 because Weave CNI doesnt support it simultaneous IPv4 IPv6 operations. 
    2. Ethernet connection to the CEC would not be supported without resolving this first.
    3. We also dont 'discover' the CEC (for mostly the same reason). Although the CEC does have programmatic IPV6 ip address based on geolocation
4. There is currently no software to support CMM to CEC via telnet
5.  The 'password' bootstraping process does not currently support the notion of a default that is not the 'singular' GLOBAL mountain default.  This would have to be expanded to accept defaults for each CEC.

## Questions

- When in the process of installation do you want or expect the ability to change the CEC passwords?
- What are your expectations regarding shipping hardware with a GLOBAL default credential?

### Potential Partial Solutions
This 'process' has a key timing constraint likely implied, that is that CEC password provisioning must occur before the installation of the system (or at least before the installation of the services that would normally be able to change passwords (SCSD, vault).

Any solution involves the physical domain, a human must force the CEC to generate the one-time-use token and read it from the LCD.   

1. Perhaps an admin on site could plug a laptop into the CEC via ethernet over ipv6.  They could set the CEC into privileged mode and follow the password change process.  They would need to record the password. 
2. admin puts the creds in  into `customizations.yaml` when the system is installed as an unencrypted sealed secret. There is a sealed secret for `REDS/MEDS/RTS`
3. Then the sealed secrets in `customizations.yaml` are encrypted.
4. `customizations.yaml`  is setup to inject the sealed secrets into the `REDS/MEDS/RTS` helm charts. The cray-service base chart is setup to create a sealed secret custom resource in k8s
5. The sealed-secrets pod, decrypts the sealed secrets custom resources and creates normal k8s secrets
6. The Vault loader jobs take the contents (default creds) of those secrets and push them into vault per CEC.
7. The discovery services would know how to fall back to searching the CEC default creds on first discovery.

Related:

1. CAST-26710
2. CAST-26712

**Comment by nieuwsma** [comment]
Some further topics for discussion related to this that came up in the COE call:

- How do blade swaps work? (SAME system) 
- How do blade swaps work across systems?
- Is there functionality to 'partially wipe or partially undiscover a blade'? This should probably do an NVME wipe which would cause the blade to get the CEC cred.
- We need to think about this process in light of 'initial install', 'adding new hardware', and 'having to change CEC creds'.  
- The 'mental model' of the creds in vault needs to expand to include CEC creds (many CECs, not just 1).
- We need to explore HOW this CEC credential gets set; for init bring up (maybe customizations.yaml?) how does it get setup on an ongoing basis?

**Reply by dmjacobsen** [reply]
In particular one of the challenges that we've run into is that because CSM records the blade NMN MAC addresses and stores them in DHCP, unless those are _removed_ before swapping a blade it won't be possible to easily bring the new blade in or move the old blade to a new location since it will keep getting the old node ids (for the NMN, but not the HSN owing to the completely different IPAM plan between the two networks).

**Comment by nieuwsma** [comment]
General feedback from COE call:

- Site was open to the idea of a 'global' password for the whole system, as long as it was 'their' default, not HPE's. 
- Site was open to the idea that a 'manual' human driven process would be necessary to set the CEC creds.
-
